### DISCUSSION TITLE
Login in BeforeAll?

### DISCUSSION THREAD

**Discussion by SpiderRam** [Tallyb/cucumber-playwright#565] [Q&A] [answered]
Disclaimer:  I am new to both cucumber and playwright, forgive any dumb questions.  😉 

I am looking for a solution to avoiding expensive repeated logins.  I have features with multiple scenarios.  The scenarios are not dependent on one another, and I would like them to have separate context, and be able to run in parallel.  This is the current behavior, but sometimes logging in to the app takes longer than what I am testing for in a single scenario. 

Is it possible to use playwright's [reuse auth](https://playwright.dev/docs/auth#reuse-signed-in-state) functionality?  I can create that storage state file by running this in the BeforeAll hook:

```ts
const page = await browser.newPage()
await page.goto('https://myapp.com/', { timeout: 10000 })
await page.locator('#email').fill(users['login_user1'].username)
await page.locator('#current-password').fill(users['login_user1'].password)
await page.locator('#password-button').click()
await page.context().storageState({ path: 'storageState.json' })
```

But there are some problems here.

1. Trying to reference this in the config appears unsuccessful.  I'm guessing that since cucumber is the test runner, not playwright, it does not know what to do with this file?
1. Different features require different users.  I am currently using tags to tell the test who to log in as.  But the BeforeAll hook does not seem able to reference tags?
1. The BeforeAll hook seems to have [other limitations](https://github.com/cucumber/cucumber-js/blob/main/docs/support_files/hooks.md#beforeall--afterall) - namely, "these methods will not have a world instance as `this`".  And that seems like this hook is not designed for the kind of action I am looking for.

Any advice is welcome.  Including just saving me from spinning my wheels on something that will never go anywhere.

**Comment by Tallyb** [comment, marked as answer]
The biggest limitation I found with cucumber is the lack of beforeFeature. BeforeAll runs only once for all tests and before runs before each **scenario**. So if you are looking for something at the feature level it does not exist. 

1. The playwright config is not used in cucumbers, exactly for the reason you mentioned. The cucumber.js file is used for all configuration needs. 
2. You can implement the same with cucumber. The before hook can take a tag and run only for those specific tags. You could also set the "logged in user" on world, and then have a default "before" (no tag)  that will login with a default user, if the current user is not set. 
3. If you need to set something in the beforeAll that will be later accessible for scenarios, you can either use a file (like above, just create it in the beforeAll) or set a global as here: https://github.com/Tallyb/cucumber-playwright/blob/master/src/support/common-hooks.ts#L27.

**Reply by SpiderRam** [reply]
Thank you!  I appreciate the info, and the speedy reply.  Also, thanks for this repo, it has been a great jumping off point.  😊

**Reply by madrane9** [reply]
> Thank you!  I appreciate the info, and the speedy reply.  Also, thanks for this repo, it has been a great jumping off point.  😊

I would be interested if you managed to implement the authentication in the common-hook. Do you have a repo?
